### PR TITLE
fix(s3): honor MetadataDirective=REPLACE for system metadata on CopyObject

### PR DESCRIPTION
**Pull request by qzhello**
## Problem

  `CopyObject` with `x-amz-metadata-directive: REPLACE` does not actually replace system metadata. The destination object inherits the source's `Content-Type`
  (and `Cache-Control`, `Content-Encoding`, `Content-Disposition`, `Content-Language`, `Expires`) regardless of what the request supplies. Only `X-Amz-Meta-*`
  user metadata is replaced.

  This silently breaks workflows that use CopyObject to fix MIME types, change CDN cache headers, or update Content-Disposition in place — a core S3 use case
  documented in the [AWS S3 CopyObject reference](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html):

  > If `MetadataDirective` is set to `REPLACE`, you can change other system-defined metadata (Content-Type, Content-Encoding, Content-Disposition,
  Cache-Control, Expires) by including the corresponding request headers.

  Reference implementation in MinIO (treated by AWS docs as a compatibility benchmark):
  - `cmd/object-handlers.go` — REPLACE extracts a fresh metadata set from the request
  - `cmd/handler-utils.go` `supportedHeaders` — the headers REPLACE rewrites

  ## Root cause

  The 4.x CopyObject refactor moved from HTTP-proxy-style copy (which transparently propagated the client request's headers to the filer PUT) to direct
  `filer_pb.Entry` construction. The new path hard-codes `Mime: entry.Attributes.Mime` from the source and never reads `Content-Type` from the request, even
  under REPLACE. `processMetadataBytes` only re-extracts `X-Amz-Meta-*` and leaves other system headers untouched.

  The normal CopyObject branch also applies `processedMetadata` to `dstEntry.Extended` via a plain upsert merge, so source-supplied managed keys (stale
  `Cache-Control`, `X-Amz-Meta-*`, `X-Amz-Tagging-*`) leak through even when the request intends to replace them. The self-copy path already handled this
  correctly via `routedMetadataReplace`.

  ## Reproduction

  ```python
  import boto3
  s3 = boto3.client("s3", endpoint_url="http://localhost:8333", ...)
  s3.put_object(Bucket="b", Key="src", Body=b"x", ContentType="image/png")
  s3.copy_object(
      Bucket="b", Key="dst",
      CopySource={"Bucket": "b", "Key": "src"},
      MetadataDirective="REPLACE",
      ContentType="text/plain",
  )
  print(s3.head_object(Bucket="b", Key="dst")["ContentType"])
  # Before: image/png  (wrong — inherited from source)
  # After:  text/plain (correct — request override honored)
  ```

  ## Changes

  - `resolveDestinationMime` derives `Attributes.Mime` from the request's `Content-Type` under REPLACE; falls back to `binary/octet-stream` when the request
  did not send one (matches MinIO's behavior in `extractMetadata`).
  - `processMetadataBytes` now also rewrites the `copyReplaceSystemHeaders` whitelist (`Cache-Control`, `Content-Encoding`, `Content-Disposition`,
  `Content-Language`, `Expires`) under REPLACE, or inherits them under COPY.
  - `isManagedCopyMetadataKey` extended to cover the same whitelist so `mergeCopyMetadata` drops stale source values.
  - The normal CopyObject branch now applies metadata via `mergeCopyMetadata`, same as the self-copy branch, so stale managed keys do not leak through on
  REPLACE.
  - Reject invalid `x-amz-metadata-directive` / `x-amz-tagging-directive` values with `ErrInvalidMetadataDirective` / `ErrInvalidTagDirective`
  (`InvalidArgument` 400) instead of silently downgrading to COPY. Mirrors MinIO error codes and messages exactly.

  ## Behavior compatibility

  | Scenario | Before | After |
  |----------|--------|-------|
  | `MetadataDirective` absent or `COPY` | inherit from source | unchanged |
  | `REPLACE` + `Content-Type` in request | source leaked | request wins |
  | `REPLACE` + no `Content-Type` in request | source leaked | `binary/octet-stream` (AWS/MinIO default) |
  | `REPLACE` + `Cache-Control` / `Content-Encoding` / `Content-Disposition` / `Content-Language` / `Expires` | source leaked | request wins (or dropped if
  request omits) |
  | `MetadataDirective=FOO` (invalid) | silently treated as COPY | 400 `InvalidArgument` |
  | `TaggingDirective=FOO` (invalid) | silently treated as COPY | 400 `InvalidArgument` |
  | Stale `X-Amz-Meta-*` / `X-Amz-Tagging-*` from source under REPLACE | leaked through | dropped |

  COPY (default) path is unchanged.

  ## Tests

  - `TestResolveDestinationMime` — Mime resolution under COPY/REPLACE + edge cases
  - `TestIsValidDirective` — accepted vs rejected directive values
  - `TestProcessMetadataBytes_ReplaceSystemHeaders` — REPLACE rewrites whitelist
  - `TestProcessMetadataBytes_CopyInheritsSystemHeaders` — COPY inherits, ignores request
  - `TestIsManagedCopyMetadataKey_CoversSystemHeaders` — whitelist registered as managed
  - `TestMergeCopyMetadata_ReplaceDropsStaleSystemHeader` — end-to-end caller flow drops stale system + user metadata
  - `TestMergeCopyMetadata_ReplaceTaggingDropsStaleTags` — end-to-end tag-REPLACE drops stale tags

  ## Related

  - AWS S3 CopyObject reference: <https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html>
  - MinIO reference implementation: <https://github.com/minio/minio/blob/master/cmd/handler-utils.go>

## Summary by CodeRabbit

* **Bug Fixes**
  * Validate metadata and tagging directives (reject invalid values with proper errors)
  * Prevent stale managed headers/tags from leaking during COPY/REPLACE; canonicalize keys case-insensitively and apply deterministic precedence
  * Fix Content-Type resolution: REPLACE uses request Content-Type or falls back to binary/octet-stream; COPY preserves source MIME

* **Tests**
  * Added comprehensive tests for directive validation, MIME resolution, metadata merging, canonical-vs-legacy key precedence, and tagging behavior

[Review Change Stack](https://app.coderabbit.ai/change-stack/seaweedfs/seaweedfs/pull/9721?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: weed/s3api/s3api_object_handlers_copy.go [hunk 1]
```diff
@@ -565,7 +609,17 @@ func isManagedCopyMetadataKey(key string) bool {
...
+		if strings.EqualFold(key, h) {
+			return true
+		}
+	}
+	// Match X-Amz-Meta-* / X-Amz-Tagging case-insensitively so legacy
+	// non-canonical keys (written by non-S3 paths or older versions) are
+	// still recognized as managed.
+	lowerKey := strings.ToLower(key)
+	return strings.HasPrefix(lowerKey, strings.ToLower(s3_constants.AmzUserMetaPrefix)) ||
+		strings.HasPrefix(lowerKey, strings.ToLower(s3_constants.AmzObjectTagging))
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/s3api/s3api_object_handlers_copy.go]: Done in 4523f72 — switched to the allocation-free `hasPrefixFold` helper (already used elsewhere in this file) rather than hardcoding lowercased literals, so the constants stay the single source of truth and there is no per-key allocation.

File: weed/s3api/s3api_object_handlers_copy.go [hunk 2]
```diff
@@ -242,7 +286,7 @@ func (s3a *S3ApiServer) CopyObjectHandler(w http.ResponseWriter, r *http.Request
 			FileSize: entry.Attributes.FileSize,
 			Mtime:    t.Unix(),
 			Crtime:   entry.Attributes.Crtime,
-			Mime:     entry.Attributes.Mime,
+			Mime:     resolveDestinationMime(r.Header, entry.Attributes.Mime, replaceMeta),
 		},
```
[the remark this answers is not in the document]
Comment (chrislusf) [reply weed/s3api/s3api_object_handlers_copy.go:298]: Fixed in 9e3ba02a. Rather than threading the Mime through `routedMetadataReplace` (PATCH_EXTENDED only carries Extended keys), `routeInPlace` is now gated on `!mimeChanged`, so a Content-Type change always takes the locked clone path, which sets `Attributes.Mime` via `resolveDestinationMime`. The routed path is used only when the Mime is unchanged. Covers the versioned self-copy case too.
